database.go

```diff
@@ -0,0 +1,272 @@
+package main
+
+import (
+	"sort"
+	"sync/atomic"
+)
+
+type Database interface {
+	BulkWrite([]*BulkMetric)
+	Open()
+	Close()
+	GetMedian() int
+}
+
+type MedianDatabase struct {
+	writeCh chan []*BulkMetric
+	quitCh  chan bool
+
+	// used to keep the left and right in sync!
+	left   []BulkMetric
+	right  []BulkMetric
+	size   int
+	median int32
+}
+
+func NewMedianDatabase() *MedianDatabase {
+	return &MedianDatabase{
+		writeCh: make(chan []*BulkMetric),
+		quitCh:  make(chan bool),
+		median:  0,
+	}
+}
+
+func (m *MedianDatabase) Open() {
+	go func() {
+		m.worker()
+	}()
+}
+
+func (m *MedianDatabase) Close() {
+	// signal the database to close itself
+	m.quitCh <- true
+
+	// wait for the database to finish processing
+	<-m.quitCh
+	close(m.quitCh)
+	close(m.writeCh)
+}
+
+func (m MedianDatabase) GetMedian() int {
+	// load the median value, ensuring that if this happens at the same
+	// time an update is happening that invalid data isn't accidentally
+	// returned
+	median := atomic.LoadInt32(&(m.median))
+
+	// by default int will correspond to the size of the current OS in go.
+	// Thus, it is safe to assume that this call will operate normally
+	// unless we are on a 16bit system (which is likely to cause much
+	// bigger problems elsewhere)
+	return int(median)
+}
+
+func (m *MedianDatabase) BulkWrite(bulkMetrics []*BulkMetric) {
+	// NOTE this is totally slow an unoptimized in every way possible; this
+	// is the quickest implementation to sort this sort of set and we want
+	// to keep it out of the critical path of `worker`.
+	sortedKeys := make([]int, 0, len(bulkMetrics))
+	keyToMetrics := make(map[int]*BulkMetric)
+
+	for _, bulkMetric := range bulkMetrics {
+		sortedKeys = append(sortedKeys, bulkMetric.Value())
+		keyToMetrics[bulkMetric.Value()] = bulkMetric
+	}
+	sort.Ints(sortedKeys)
+
+	// TODO implement a nicer type around `[]*BulkMetric` which implements
+	// the Sort interface
+	for i, key := range sortedKeys {
+		bulkMetrics[i] = keyToMetrics[key]
+	}
+
+	m.writeCh <- bulkMetrics
+}
+
+func (m *MedianDatabase) worker() {
+	// in order to efficiently sort and maintain a median for a high volume
+	// of data we effectively organize data into two arrays, the left and
+	// right side. In normal, balanced circumstances, when len(left) ==
+	// len(right) we can guarantee that the median is either left[-1] or
+	// average(left[-1], right[0]).
+	left := make([]*BulkMetric, 0, 1000)
```

Comment:
**note** this method is the bread and butter of the metrics storage engine and "on the fly" calculation engine. It probably could use being pulled out from just being a self contained `goroutine` later on to make testing, benchmarking easier etc.

It probably makes sense to pull out any knowledge of the `median` bits here and add some sort of mutex around the head of the `right` and the tail of `left` themselves. As is stands, all memory is contained in this single goroutine so as to prevent any memory sharing (apart from the single `m.median` value which is protected by atomic read methods above).